rust_src/src/buffers.rs

```diff
@@ -434,4 +474,34 @@ pub extern "C" fn nsberror(spec: Lisp_Object) -> ! {
     }
 }
 
+/// These functions are for debugging overlays.
+
+/// Return a pair of lists giving all the overlays of the current buffer.
+/// The car has all the overlays before the overlay center;
+/// the cdr has all the overlays after the overlay center.
+/// Recentering overlays moves overlays between these lists.
+/// The lists you get are copies, so that changing them has no effect.
+/// However, the overlays you get are the real objects that the buffer uses.
+#[lisp_fn]
+pub fn overlay_lists() -> LispObject {
+    let list_overlays = |ol: Option<LispOverlayRef>| -> LispObject {
+        let mut ol_list = Qnil;
+
+        if let Some(o) = ol {
+            let mut it = o.iter();
+
+            while let Some(n) = it.next() {
+                ol_list = unsafe { Fcons(n.as_lisp_obj().to_raw(), ol_list) };
+            }
+        }
+
+        LispObject::from_raw(ol_list)
+    };
+
```

Comment:
This closure could also be written:

    let list_overlays = |ol: Option<LispOverlayRef>| -> LispObject {
        if let Some(o) = ol { // or use match...
            let ol_list = o.iter().fold(Qnil, |(accum, n)| unsafe { Fcons(n.as_lisp_obj().to_raw(), accum) });
            LispObject::from_raw(ol_list)
        } else {
            LispObject::constant_nil()
        }
    }

Reply:
Nice

Reply:
```
error[E0619]: the type of this value must be known in this context
   --> /home/fredbert/tmp/git/remacs/rust_src/src/buffers.rs:490:23
    |
490 |                 Fcons(n.as_lisp_obj().to_raw(), accum)
    |                       ^^^^^^^^^^^^^^^
```

Hmmm. I'm not sure how to solve this elegantly.

Reply:
My mistake, I slipped into Haskell style :-)

The lambda does not take a tuple. It should be `|accum, n|` not `|(accum, n)|`.

Reply:
pub fn overlay_lists() -> LispObject {
        let list_overlays = |ol: LispOverlayRef| -> LispObject {
            let ol_list = ol.iter().fold(Qnil, |accum, n| unsafe {
                Fcons(n.as_lisp_obj().to_raw(), accum)
            });
            LispObject::from_raw(ol_list)
        };
    
        let cur_buf = ThreadState::current_buffer();
        let before = cur_buf
            .overlays_before()
            .map_or_else(|| LispObject::constant_nil(), &list_overlays);
        let after = cur_buf
            .overlays_after()
            .map_or_else(|| LispObject::constant_nil(), &list_overlays);
        unsafe { LispObject::from_raw(Fcons(Fnreverse(before.to_raw()), Fnreverse(after.to_raw()))) }
    }